auth: Add JWT-based user API key fetch.

mateuszmandera opened the pull request:
This is just https://github.com/zulip/zulip/pull/22955 but Github doesn't allow me to push my tweaks to the PR, so pushing it here for now. 

@atoppi FYI

I rebased this on top of `main` resolving conflicts and added a commit to extract the very-similar logic between the new endpoint and remote_user_jwt. It makes some changed to remote_user_jwt (elaborated in the commit) but I think basically noone relies on that API anyway? Not sure

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -152,6 +152,23 @@
     # "zproject.backends.GenericOpenIdConnectBackend",  # Generic OIDC integration, setup below
 )
 
+## JWT-based users login and API keys fetching.
+##
+## If you want to login users or fetch API keys via JWT in Zulip,
+## you need to set the secret key and algorithm to use to validate
+## JWT tokens received.
+# JWT_AUTH_KEYS: Dict[str, Any] = {
+#     # Subdomain for which you want to enable the service.
+#     "zulip": {
+#         # Shared secret key used to validate jwt tokens. This key
+#         # needs to be securely, randomly generated.
+#         "key": "xxxxxx",
+#         # Algorithm with which the jwt token are signed.
+#         "algorithms": ["HS256"],
+#     }
+# }
```
I amended the original commit to comment out these lines, because otherwise this config was going to be enabled by default for all new installations

atoppi replied:
makes sense

mateuszmandera commented:
@timabbott I think this is ready for your review now.

atoppi commented:
@mateuszmandera thanks for porting the original patch to the `main` branch.
Can't comment about the changes to `remote_user_jwt`, but for the rest it looks good to me!

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -152,6 +152,23 @@
+#     # Subdomain for which you want to enable the service.
+#     "zulip": {
+#         # Shared secret key used to validate jwt tokens. This key
+#         # needs to be securely, randomly generated.
```
It would be good to comment more on the parameters for said generation -- e.g. can we say anything about the entropy requirements for this to be secure?

mateuszmandera replied:
Amended this paragraph to mention that:
```
#         # Shared secret key used to validate jwt tokens. This key
#         # needs to be securely, randomly generated. Note that if you're
#         # using the default HS256 algorithm, per RFC 7518, the key needs
#         # to be at least 256 bits long.
```

would it be better to phrase this in terms of bits of entropy rather than key length? (I went with how the RFC talks about it). Also ref https://auth0.com/blog/brute-forcing-hs256-is-possible-the-importance-of-using-strong-keys-to-sign-jwts/ which mentions this requirement, though they say they use 512 bits.

timabbott replied:
Yeah, bits of entropy is the right metric to talk about.

mateuszmandera replied:
Tweaked

timabbott commented on zproject/urls.py:
```diff
@@ -723,6 +724,12 @@
+#  and the details for a single user.
+urls += [
+    path("jwt/fetch_api_key", jwt_fetch_api_key),
+]
```
I think this should probably be in the API URL namespace, like `api/v1/jwt/fetch_api_key`, perhaps?

Also, I'd move this after `Python Social Auth` in the file?

mateuszmandera replied:
> Also, I'd move this after Python Social Auth in the file?

Yeah seems better, moved it and put it in  `/api/v1`.

atoppi replied:
for what it's worth, the original namespace was decided by looking at `json/fetch_api_key`
```
# Desktop-specific authentication URLs
urls += [
    rest_path("json/fetch_api_key", POST=json_fetch_api_key),
]
```

timabbott commented on zerver/views/auth.py:
```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+    )
+    if user_profile is not None:
+        assert isinstance(user_profile, UserProfile)
+
```
Is this if/assert block required?

mateuszmandera replied:
I think so, deleting it causes mypy errors - we introduced these asserts after `authenticate` calls in 4c6f2ae7bee23980c6bc0d0ac2ce144c75aa14ee Though this is made nicer due to the change discussed in https://github.com/zulip/zulip/pull/24199#discussion_r1089571167

timabbott commented on zerver/views/auth.py:
```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+    api_key = api_key = process_api_key_fetch_authenticate_result(
+        request, user_profile, return_data
+    )
+    assert user_profile is not None
```
I don't understand this bit -- I think we're likely missing an exit path for this function if `user_profile is None` -- i.e. a failed authentication.

timabbott replied:
Oh, I see, `process_api_key_fetch_authenticate_result` raises an exception in that case. That deserves a comment at least; there's probably a fancy type assertion we could declare on `process_api_key_fetch_authenticate_result` as well.

But maybe a better plan would be to have a function that we call if `user_profile` is None called `process_api_key_fetch_authenticate_failure`, and have the logic to raise the appropriate exception be there; or maybe even `raise get_api_key_fetch_authenticaiton_exception(return_data)` (with a `return` rather than `raise` as what that function returns).

timabbott replied:
Yeah, that looks good.

timabbott commented on zerver/views/auth.py:
```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+        key = settings.JWT_AUTH_KEYS[realm.subdomain]["key"]
+        algorithms = settings.JWT_AUTH_KEYS[realm.subdomain]["algorithms"]
+    except KeyError:
+        raise JsonableError(_("Auth key for this subdomain not found"))
```
Let's make this more appropriate for the end user. For the sysadmin, something like "Subdomain {subdomain} does not have JWT_AUTH_KEYS configured." would be appropriate exception text. But I think that belongs as a `logging.info` or if wanted that.

I think a reasonable `JsonablError` here would be something like "JWT authentication is not enabled for this organization."

mateuszmandera replied:
Improved the error. Yeah the logging is probably not needed since it doesn't reveal anything that the admin shouldn't have seen already by making a test auth request when configuring this stuff?

timabbott replied:
Yeah, I think I agree.

timabbott commented on zerver/views/auth.py:
```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+        raise JsonableError(_("Auth key for this subdomain not found"))
+
+    try:
+        json_web_token = request.POST["json_web_token"]
```
We should be using the parameter from REQ here.

Also, is `json_web_token` a reasonably standard name for a JWT token parameter? I feel like I'd expect something more like `jwt` or `token` or something, but have no experience with JWT personally.

mateuszmandera replied:
Hmm yeah a generic `token` seems more standard, tweaked that + added REQ + `@require_post`

timabbott commented on zerver/views/auth.py:
```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+
+    remote_email = payload.get("email", None)
+    if remote_email is None or not remote_email:
+        raise JsonableError(_("No email specified in JSON web token claims"))
```
Should we run an email address validator here? I think the answer is no, since we only do that in account creation normally, and this is not an account creation code path.

I'm also not sure we need the `or not remote_email` code here; `""` seems equally broken as an email address as say, `" "`, so unless there's a reason we'd expect that to be a common failure mode, I feel like the `Invalid email` exception that `authenticate` is going to be just as useful?

mateuszmandera replied:
> Should we run an email address validator here? I think the answer is no, since we only do that in account creation normally, and this is not an account creation code path.

Yeah that was my thinking here too

> I'm also not sure we need the `or not remote_email` code here; `""` seems equally broken as an email address as say, `" "`, so unless there's a reason we'd expect that to be a common failure mode, I feel like the `Invalid email` exception that `authenticate` is going to be just as useful? 

I'm guessing that the idea was that the empty string is failing to submit a value as opposed to something malformed like `" "` - but the `Invalid email` treatment makes sense to me too. Perhaps  @atoppi can you comment your idea regarding this?

atoppi replied:
The `or not remote_email` part can be removed, it was just an extra check that we do not need anymore.

mateuszmandera replied:
Cool, pushed a commit cleaning that up.

timabbott commented on zproject/prod_settings_template.py:
```diff
@@ -152,6 +152,23 @@
+##
+## If you want to login users or fetch API keys via JWT in Zulip,
+## you need to set the secret key and algorithm to use to validate
+## JWT tokens received.
```
We should probably rework this to have a bit more documentation in `docs/authentication-methods.md`, with this just being a stub linking to there?

timabbott replied:
Also, this should definitely not be the first block in this file; let's move it to be the last of the authentication blocks in this file.

mateuszmandera replied:
I wrote up a brief doc in `authentication-methods.md` and pushed.

timabbott commented:
Thanks for picking this up @mateuszmandera! I posted a bunch of tactical comments; I also haven't read the tests, can you comment on to what extent you feel satisfied with them?

mateuszmandera commented:
> I also haven't read the tests, can you comment on to what extent you feel satisfied with them?

They look good to me.

@timabbott I made the various tweaks, so should be ready for more review. Left https://github.com/zulip/zulip/pull/24199#discussion_r1089573961 unresolved for now, waiting for feedback from @atoppi on that error handling since he's going to be more familiar with their use of this endpoint and potential client errors

mateuszmandera commented on zproject/prod_settings_template.py:
```diff
@@ -534,6 +535,26 @@
+#         # needs to be securely, randomly generated. Note that if you're
+#         # using the default HS256 algorithm, per RFC 7518, the key needs
+#         # to be at least 256 bits long.
+#         "key": "xxxxxx",
```
Hmm, also thinking now that this should probably be moved to be read from `zulip-secrets.conf` instead?

alexamirante replied:
This would break backward compatibility, though, as the `JWT_AUTH_KEYS` dictionary was already in place and used by the `/json/fetch_api_key` API. Unless the upgrade scripts (e.g., `upgrade-zulip-from-git`) will take care of moving the key from `settings.py` to `zulip-secrets.conf`?

timabbott replied:
Yes, that's correct.

mateuszmandera replied:
Changed this to use `get_secrets`.

@alexamirante This won't break compatibility, it'll only change where the key is fetched from for new installation. An old installation that has set this up will have their `/etc/zulip/settings.py` with `JWT_AUTH_KEYS` just specifying `key` there in the file, so it won't be affected. (That's because `prod_settings_template.py` simply becomes `/etc/zulip/settings.py` when you're doing a new installation)

mateuszmandera commented:
@timabbott https://github.com/zulip/zulip/pull/24199#discussion_r1089573961 is now cleaned up and the secret key moved to `zulip-secrets` as discussed above.

timabbott commented:
Cool, how do you think we should plan to merge this? I think most of the changes seem best squashed into the original commit with a `Co-authored-by`, but perhaps there's some refactoring that should be moved to the base as prep commits?

mateuszmandera commented:
> Cool, how do you think we should plan to merge this? I think most of the changes seem best squashed into the original commit with a `Co-authored-by`, but perhaps there's some refactoring that should be moved to the base as prep commits?

@timabbott  Yeah that way of merging will be cleanest - I restructured the commit so that basically everything affecting other code is in prep commits and then we just add a single, reasonably clean `auth: Add JWT-based user API key fetch.` commit with `Co-authored-by` (meaning also all the tweaks have been squashed into it).

Also tweaked the doc phrasing about the secret key to talk about bits of entropy, as discussed above.

timabbott commented:
I made a few tweaks:
* Removing the unused `return_data` parameter from `process_api_key_fetch_authenticate_result`
* Small edits to the documentation, both in `docs/` and `prod_settings_template.py` for English writing polish and hopefully clarity. @atoppi I'd greatly appreciate a thoughtful review especially on the `docs/` part of this on whether there's anything we could explain better.
* Added "Upgrade notes" documenting the change in the existing JWT support.

With those changes, I think this is good to merge pending feedback from the other folks who've worked on this. I opened #24278 for a follow-up on documentation that would be part of making this feature a polished official feature. I imagine we'll also want to think through how to advertise that this exists -- feedback on where you'd go looking for it would be appreciated!

atoppi commented:
> * Small edits to the documentation, both in `docs/` and `prod_settings_template.py` for English writing polish and hopefully clarity. @atoppi I'd greatly appreciate a thoughtful review especially on the `docs/` part of this on whether there's anything we could explain better or should really mention. Preview here: https://zulip--24199.org.readthedocs.build/en/24199/production/authentication-methods.html#jwt

@timabbott the docs look good to me. It is clear and explains well how we are currently using the feature.

mateuszmandera commented on docs/overview/changelog.md:
```diff
@@ -21,6 +21,10 @@ log][commit-log] for an up-to-date list of all changes.
 - Removed the `application_server.no_serve_uploads` setting in
   `/etc/zulip/zulip.conf`, as all uploads requests go through Zulip now.
+- Installations using the previously undocumented [JWT authentication
+  feature](../authentication/authentication-methods.md#jwt) will need
```
Should be `../production/authentication-methods.md`, I'll tweak that to fix CI

mateuszmandera commented:
Made tiny tweaks, including fixing CI:
```
diff --git a/docs/overview/changelog.md b/docs/overview/changelog.md
index 2e6adc3c9d..105b303c59 100644
--- a/docs/overview/changelog.md
+++ b/docs/overview/changelog.md
@@ -22,7 +22,7 @@ log][commit-log] for an up-to-date list of all changes.
 - Removed the `application_server.no_serve_uploads` setting in
   `/etc/zulip/zulip.conf`, as all uploads requests go through Zulip now.
 - Installations using the previously undocumented [JWT authentication
-  feature](../authentication/authentication-methods.md#jwt) will need
+  feature](../production/authentication-methods.md#jwt) will need
   to make minor adjustments in the format of JWT requests; see the
   documentation for details on the new format.
 
diff --git a/docs/production/authentication-methods.md b/docs/production/authentication-methods.md
index 076d4f8ca1..35abc4196f 100644
--- a/docs/production/authentication-methods.md
+++ b/docs/production/authentication-methods.md
@@ -870,7 +870,7 @@ Zulip supports using JSON Web Tokens (JWT) authentication in two ways:
 
 1. Obtaining a logged in session by making a POST request to
    `/accounts/login/jwt/`. This allows a separate application to
-   integrate by Zulip via having a button that directly takes the user
+   integrate with Zulip via having a button that directly takes the user
    to Zulip and logs them in.
 2. Fetching a user's API key by making a POST request to
    `/api/v1/jwt/fetch_api_key`. This allows a separate application to
diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
index 122de397ca..4bcf7e91c2 100644
--- a/templates/zerver/api/changelog.md
+++ b/templates/zerver/api/changelog.md
@@ -24,8 +24,9 @@ format used by the Zulip server that they are interacting with.
 
 * [`POST /api/v1/jwt/fetch_api_key`]: New API endpoint to fetch API
   keys using JSON Web Token (JWT) authentication.
-* [`accounts/login/jwt/`]: Adjusted format of requests optional
-  previously undocumented JWT authentication log in support.
+* [`accounts/login/jwt/`]: Adjusted format of requests to this
+  previously undocumented, optional endpoint for
+  JWT authentication log in support.
```

Resolved merge conflicts as well.

timabbott commented:
Merged, thanks @atoppi and @mateuszmandera!
